wrappers/flutter/lib/src/utility/resource_lease_manager.dart

```diff
@@ -0,0 +1,286 @@
+import 'dart:async';
+import 'dart:developer' as dev;
+import 'dart:collection';
+
+/// Asynchronous factory invoked when a lease becomes active.
+///
+/// - Called when the lease reaches the front of the queue.
+/// - Should return a ready‑to‑use resource (e.g. an initialized camera controller).
+/// - May be wrapped by the manager's `createTimeout`.
+typedef ResourceFactory<T> = Future<T> Function();
+
+/// Function that releases the resource after the lease ends.
+///
+/// - Invoked exactly once for each created resource.
+/// - Must fully and safely release the resource (e.g. `dispose`).
+/// - May be wrapped by the manager's `releaseTimeout`.
+typedef ResourceReleaser<T> = Future<void> Function(T resource);
+
+/// Logger interface — allows injecting a custom logger.
+typedef ResourceLogger = void Function(String message, Object? error, StackTrace? stackTrace);
+
+/// Manager that provides exclusive (sequential) access to a resource via leases.
+///
+/// How it works:
+/// - Only one lease can be active at a time; others wait in a FIFO queue.
+/// - When a lease becomes active, the manager calls the `ResourceFactory`, then
+///   completes `lease.resource`. Use the resource only within that lease window.
+/// - When finished, call `lease.release()` / `lease.releaseWaiting()`; the manager
+///   calls the `ResourceReleaser` and moves to the next lease.
+/// - If a lease is released before the resource is created, creation is skipped.
+/// - Optional `createTimeout` / `releaseTimeout` protect against hangs.
+///
+/// Typical use: serialize access to `CameraController` / `MobileScannerController`.
+/// Prevents concurrent camera use and simplifies the resource life cycle.
+class ResourceLeaseManager {
+  ResourceLeaseManager({
+    ResourceLogger? logger,
+    this.createTimeout,
+    this.releaseTimeout,
+    this.maxQueueLength,
+  }) : _logger = logger;
+
+  final _leases = ListQueue<ResourceLease>();
+  final ResourceLogger? _logger;
+  final Duration? createTimeout;
+  final Duration? releaseTimeout;
+  final int? maxQueueLength;
+
+  /// Number of items in the queue (active + pending).
+  int get queueLength => _leases.length;
+
+  /// Whether there is an active or pending lease.
+  bool get hasActiveLease => _leases.isNotEmpty;
+
+  void _log(String component, String message, {Object? error, StackTrace? stackTrace, int level = 800}) {
+    final formatted = '[$component] $message';
+    _logger?.call(formatted, error, stackTrace);
+    dev.log(message, name: component, level: level, error: error, stackTrace: stackTrace);
```

Comment:
nie powinniśmy tego jakos kDebug owrapować? Analogicznie w pozostałych miejscach
```suggestion
    if (kDebug) dev.log(message, name: component, level: level, error: error, stackTrace: stackTrace);
```

Reply:
Te logi nie będą widoczne w release, a będą widoczne w debug i profile. 
Metoda `log` z `dart:developer` właśnie tak działa. Logi można podejrzeć w DevToolsach